OneCommunityGlobal/HGNRest
# Fix weekly badge assignment tasks do not update the badge earned date collection

# Description
**Fix the badge  earned date not being updated in the weekly badge assignment task**
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/c41953a1-77b0-466e-952d-6f0c1af5f991)

**Fix the issue where the last modified date mismatches with the last earned date in the frontend.**
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/c1a773e7-5140-4d39-8db0-9cf142abe0b1)

## Related PRS (if any):
This backend PR is related to the #[1671](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1761) frontend PR.
To test this backend PR you need to check out the #1671 frontend PR.
…

## Main changes explained:
- Update file userHelper.js to include new code to process existing earned date records when the badge count is updated
- Update file userHelper.js to fix all linting issues


## How to test:
1. check into the current branch
2. do `npm install` and `...` to run this PR locally
3. Clear site data/cache
4. log in as an admin/owner user
5. to test the last modified date mismatches with the last earned date fix (no backend modification needded):
      - go to View Profile link Or Others -> Badge Management
      - add a new badge to a user and verify the newly added badge's last modified date is the same as the latest earned date
      - update an existing badge and verify the badge's last modified date is the same as the latest earned date
6. to test the weekly badge assignment fix:
      - go to the userProfileJobs.js file in the backend
      - **modify the line**  '1 0 * * *' to '* * * * *' so the cron job runs every minute
      - **comment out all scheduled tasks but userhelper.awardNewBadges()** -> so you don't run irrelevant jobs
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/1e220da3-3742-46a1-957d-8bf4836c4da2)
      - go to the userHelper.js file in the backend
      - **modify the awardNewBadges()** function to run tasks on a specific user (an account to which you have access would be the best) -> so you don't run the badge award function to all users
       - We are doing unit testing on the modified function here: Comment out 1494-1515. 
      ![image](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/69882989/376d03ca-d1bd-4eea-9314-113059dd01a1)
      - code snippet

      ```
      /*
       * the changeBadgeCount takes 3 arguments userId, badgeId, and newBadgeCount. 
       * The provided code snippet retrieves the user by user id and updates the existing badge records. 
       * To run the sample testing code, all you need is to change the input 'userId ' to the actual user id you prefer.
      */
      
             // retrieves user
             const userId = '65500b658e0b2922b80d5f9f'; // replace this to your user id
             const user = await userProfile
             .findOne({ _id: userId })
             .populate('badgeCollection.badge');
             // unit testing for changeBadgeCount
             for (let badgeItem of user.badgeCollection){
               await changeBadgeCount(userId, badgeItem.badge._id, badgeItem.count + 1); // or any number you like
             }
      
             if (cache.hasCache(`user-${userId}`)) {
                  cache.removeCache(`user-${userId}`);
             }
     ```

     - verify that **all badges count increase by 1 or any number you specified in the previous code** and **the last modified date should match the latest earned date**. Refresh the frontend page after seeing 'Awarding' in the backend console.
      ![image](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/474c0c0e-c280-40a3-987a-a8d5c55fd3b4)


## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/e7f2f0c0-1fc0-419c-975b-e7955e1ebda1




## Note:
Include the information the reviewers need to know.
